### DISCUSSION TITLE
Read USB port on Raspberry Pi Pico with alternative to pyserial

### DISCUSSION THREAD

**Discussion by hanssens** [micropython/micropython#10422] [Using MicroPython]
Hi, I'm trying to read out a USB-connected device hooked onto my Raspberry Pi Pico, which in CPython is quite easily done using the [pyserial lib](https://pyserial.readthedocs.io/en/latest/):

```python
ser = serial.Serial()
ser.baudrate = 115200
ser.bytesize = serial.EIGHTBITS
ser.parity = serial.PARITY_NONE
ser.stopbits = serial.STOPBITS_ONE
ser.xonxoff = 0
ser.rtscts = 0
ser.timeout = 12

# configure usb port
ser.port = "/dev/ttyUSB0"
ser.close()

try:
    ser.open()

    raw_msg = ser.readline()
    decoded = raw_msg.decode('ascii').strip()
except:
    sys.exit(f'Error on {ser.name}')
finally:
    ser.close
```

Considering pyserial isn't available in micropython, what can I dive into to achieve the same? I.e. read out data from a USB connected device. 

Please note that I'm new to micropython and am looking to both [the docs](https://micropython.org/) and the source, but can't find a good example for this use case (USB) yet. 

Thanks in advance.

**Comment by karfas** [comment]
USB Host and Device functions for micropython-controlled boards are work in progress, I think (see e.g. #9356 and #9497).

Your Linux-based example works mostly because the real communication to the USB device runs over some more or less complex drivers, all hidden behind the automagically created `/dev/ttyUSB0` virtual serial device.

**Reply by jean-claudeF** [reply]
Could this eventually help you? :
http://staff.ltam.lu/feljc/electronics/uPython/Pico_communication.pdf

**Reply by jimmo** [reply]
Exactly what @karfas said -- you need the following things to be implemented:
 - USB host mode support (in progress)
 - CDC/ACM device class support (definitely something that would be worthwhile once host mode is implemented)

The result of this would be something that works a bit like a UART instance (similar to pyb.USB_VCP for device-mode CDC/ACM).

**Reply by badgermedia** [reply]
Is there an approximate timeline for USB Host support? I am looking for exactly this functionality along with the ability to support serial connections to USB peripherals (also MP based products). Many thanks
